relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -15,79 +15,63 @@ use crate::envelope::{Envelope, ItemType};
 /// Checks whether an event should be kept or removed by dynamic sampling.
 pub fn should_keep_event(
     sampling_context: Option<&DynamicSamplingContext>,
-    event: &Event,
+    event: Option<&Event>,
     ip_addr: Option<IpAddr>,
     project_state: &ProjectState,
+    trace_root_project_state: Option<&ProjectState>,
     processing_enabled: bool,
 ) -> SamplingResult {
-    let sampling_config = match &project_state.config.dynamic_sampling {
-        // without config there is not enough info to make up my mind
-        None => return SamplingResult::NoDecision,
-        Some(config) => config,
-    };
-
-    // when we have unsupported rules disable sampling for non processing relays
-    if !processing_enabled && sampling_config.has_unsupported_rules() {
-        return SamplingResult::Keep;
+    let mut matching_rule = None;
+
+    if let (Some(sampling_context), Some(trace_root_project_state)) =
+        (sampling_context, trace_root_project_state)
+    {
+        if let Some(ref sampling_config) = trace_root_project_state.config.dynamic_sampling {
+            // when we have unsupported rules disable sampling for non processing relays
+            if !processing_enabled && sampling_config.has_unsupported_rules() {
+                return SamplingResult::Keep;
```

Comment:
This cannot return, as you will no longer check for individual event rules below. The checks should be independent from each other, though.